### PR TITLE
Bug Fixes and Race Controls

### PR DESCRIPTION
**Pull request by Jakllp**
- Fixed some bugs (see commit notes)
- Race Controls HAD some problems
- Can now drive when standing on edge of block

### CODE REVIEW

**Comment by Jakllp** [comment]
(ignore the travis stuff I was trying some things and forgot to make it a branch...)

File: src/main/java/com/useful/ucars/MotionManager.java [hunk 1]
```diff
@@ -111,6 +111,12 @@ public static void move(Player player, float f, float s, boolean jumping) { // '
 		}
 		
 		boolean inAir = car.getLocation().clone().add(0, -1, 0).getBlock().isEmpty();
+		if(!car.getLocation().clone().add(0.5, -1, 0).getBlock().isEmpty() || !car.getLocation().clone().add(-0.5, -1, 0).getBlock().isEmpty() ||
+				!car.getLocation().clone().add(0, -1, 0.5).getBlock().isEmpty() || !car.getLocation().clone().add(0, -1, -0.5).getBlock().isEmpty() ||
+				!car.getLocation().clone().add(0.5, -1, 0.5).getBlock().isEmpty() || !car.getLocation().clone().add(0.5, -1, -0.5).getBlock().isEmpty() ||
+				!car.getLocation().clone().add(-0.5, -1, 0.5).getBlock().isEmpty() || !car.getLocation().clone().add(-0.5, -1, -0.5).getBlock().isEmpty()) { //Check area under car
```
Comment (storm345) [inline src/main/java/com/useful/ucars/MotionManager.java:117]: Maybe should be using half the entity width instead of 0.5?

File: src/main/java/com/useful/ucars/uCarsListener.java [hunk 2]
```diff
@@ -291,6 +292,9 @@ public boolean isACar(Entity cart) {
 		if((cart.getType().toString().toUpperCase().contains("MINECART") && cart.getType().toString().length() > 8) || cart.getType().toString().toUpperCase().contains("BOAT")) {
 			return false; //Not a car but a Minecart_Something (only useful when derailed)
 		}
+		if(cart instanceof Animals) {
+			return false;	//Animals are not cars...
```
Comment (storm345) [inline src/main/java/com/useful/ucars/uCarsListener.java:296]: Why is this needed?

File: src/main/java/com/useful/ucars/uCarsListener.java [hunk 3]
```diff
@@ -553,22 +560,23 @@ public void tickCalcsAndLegacy(VehicleUpdateEvent event) {
...
 			if(UEntityMeta.hasMetadata(vehicle, "car.vec")){ //If it has the 'car.vec' meta, we need to use RACE CONTROLS on this vehicle
 				ucarUpdateEvent evt = (ucarUpdateEvent) UEntityMeta.getMetadata(vehicle, "car.vec").get(0).value(); //Handle the update event (Called here not directly because otherwise ppl with a better connection fire more control events and move marginally faster)
 				evt.player = ((Player)passenger); //Set the player (in the car) onto the event so it can be handled by uCarUpdate handlers
 				evt.incrementRead(); //Register that the control input update has been executed (So if no new control input event within 2 ticks; we know to stop the car)
 				UEntityMeta.removeMetadata(vehicle, "car.vec"); //Update the 'car.vec' metadata with an otherwise identical event; but without the player object attached
-				ucarUpdateEvent et = new ucarUpdateEvent(vehicle, evt.getTravelVector().clone(), null, evt.getDir()); //Clone of the other event, except no player object attached
+				ucarUpdateEvent et = new ucarUpdateEvent(vehicle, evt.getTravelVector().clone(), evt.getPlayer(), evt.getDir()); //Clone of the other event, except no player object attached
```
Comment (storm345) [inline src/main/java/com/useful/ucars/uCarsListener.java:569]: The player shouldn't be attached to the event here as I think that is a potential memory leak? (It was null before I think so that a car in the map isn't keeping player objects alive that ought to have been cleaned up.)

File: src/main/java/com/useful/ucars/uCarsListener.java [hunk 4]
```diff
@@ -1275,20 +1293,13 @@ public void run() {
...
-																		// step
-																		// block
-																		// and
-																		// trying
-																		// to
-																		// jump
+				if (carBlock.name().toLowerCase().contains("slab")) { // In a slab block
 					calculated = true;
-					y = 0.6;
+					y = 1.5;
```
Comment (storm345) [inline src/main/java/com/useful/ucars/uCarsListener.java:1302]: Why 1.5 here?

**Review by storm345 — CHANGES_REQUESTED** [review]
Mostly looks good, just a couple questions and points (see the comments on the code)

**Comment by storm345** [comment]
If you can, it'd be good to cleanup the travis commits (Squash or similar)

File: src/main/java/com/useful/ucars/uCarsListener.java (hunk 2 above)
Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:296]: Animal Vehicles like horses, pigs or striders were handled as cars when they were on roadblocks/the entire time when roadblocks were turned off

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:296]: Ah okay makes sense

File: src/main/java/com/useful/ucars/uCarsListener.java (hunk 3 above)
Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:569]: I think I added that for testing and forgot to remove it

File: src/main/java/com/useful/ucars/uCarsListener.java (hunk 4 above)
Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1302]: Will double check but I believe one was stuck on that slab otherwise

File: src/main/java/com/useful/ucars/MotionManager.java (hunk 1 above)
Comment (Jakllp) [reply src/main/java/com/useful/ucars/MotionManager.java:117]: Good point, done - commit coming soon
